### DISCUSSION TITLE
How to diagnose problems with steps.RemoveDirectory?

### DISCUSSION THREAD

**Discussion by sound-fx** [buildbot/buildbot#6347] [Q&A] [answered]
My builders, on both Windows and Linux, have trouble using steps.RemoveDirectory.  What can I do to get diagnostic information about the failure?

The twistd.log shows no problem running the step.  For example,
```

...
2022-01-27 14:50:04-0800 [Broker,2,10.20.0.19] ping finished: success
2022-01-27 14:50:04-0800 [-] releaseLocks(RemoveDirectory(name=Interpolate('Clean_%(prop:cmake_fortran)s_x64_build_folder'), dir='../build_out', haltOnFailure=False, flunkOnFailure=False, description=Interpolate('Cleaning the ../build_out directory for the %(prop:cmake_fortran)s %(prop:cmake_threads)s-threaded %(prop:lib_linking)s-library configuration of GPSII'))): []
2022-01-27 14:50:04-0800 [-] acquireLocks(build <Build gpsii_es_ifort_centos7_static_parallel_release number:6 results:success>, locks [(<MasterLock(ifort_license, 1)>, <buildbot.locks.LockAccess object at 0x7f7d8498d048>)])
...
```

but the folder, and at least some of its contents, don't actually get deleted, causing problems for subsequent steps.

How to diagnose what could be going wrong?  Is there a setting to get more details about that step -- what command gets executed and what the status code is and the error message, if any?

I don't think it's a permissions issue, as if I replace the steps.RemoveDirectory with an equivalent steps.ShellCommand, the builds seem to all work fine.  For example, on Windows, the command looks like
```
                command       = [ 'rmdir',
                                  '/s',
                                  '/q',
                                  r'..\build_out',
                                ],
```
But I'd prefer to use steps.RemoveDirectory so I don't have to have separate, customized steps for Linux and Windows.

**Comment by p12tic** [comment, marked as answer]
@sound-fx On Windows we use a python function to remove files line by line - `rmDirRecursive` from `worker/buildbot_worker/commands/utils.py`. Would it be possible for you to extract the function to a separate Python script, call it and see if it successfully removes the directory or not?

**Reply by sound-fx** [reply]
Thank you very much for the suggestion, @p12tic .

Using your suggestion, I traced the problem on Windows.  I found that some of the intermediate build files have long path names, exceeding 260 characters,.  It seems that steps.RemoveDirectory fails to remove these long-named files, while the steps.ShellCommand with 'rmdir' succeeds.

I managed to shorten the name of one of the directories just enough to bring the path names under 260 characters, and now steps.RemoveDirectory also succeeds.

I hope that the problem on Linux will have a similar diagnosis and cure, but I haven't yet had a chance to confirm that.

**Comment by sound-fx** [comment]
Thank you very much for the suggestion, @p12tic .  

Using your suggestion, I found that on Windows, some of the intermediate build files have long path names, exceeding 260 characters. It seems that on Windows, steps.RemoveDirectory fails to delete these long-name intermediate build files, while the steps.ShellCommand with 'rmdir' succeeds.

I managed to shorten one of the directory names just enough that now steps.RemoveDirectory works there once again.

I have some hope that a similar diagnosis will account for the problem on Linux as well, but I haven't had a chance to confirm that yet.
